ruff_lsp/server.py

```diff
@@ -186,38 +358,137 @@ class VersionModified(NamedTuple):
+    diagnostics = await _lint_document_impl(document)
+
+    # Publish diagnostics for each cell.
+    for cell_idx, diagnostics in _group_diagnostics_by_cell(diagnostics).items():
+        LSP_SERVER.publish_diagnostics(
+            # The cell indices are 1-based in Ruff.
+            params.notebook_document.cells[cell_idx - 1].document,
+            diagnostics,
+        )
+
+
+@LSP_SERVER.feature(NOTEBOOK_DOCUMENT_DID_CLOSE)
+def did_close_notebook(params: DidCloseNotebookDocumentParams) -> None:
+    """LSP handler for notebookDocument/didClose request."""
+    # Publishing empty diagnostics to clear the entries for all the cells in this
+    # Notebook Document.
+    for cell_text_document in params.cell_text_documents:
+        LSP_SERVER.publish_diagnostics(cell_text_document.uri, [])
+
+
+@LSP_SERVER.feature(NOTEBOOK_DOCUMENT_DID_SAVE)
+async def did_save_notebook(params: DidSaveNotebookDocumentParams) -> None:
+    """LSP handler for notebookDocument/didSave request."""
+    await _did_change_or_save_notebook(
+        params.notebook_document.uri, run_types=[Run.OnSave, Run.OnType]
+    )
+
+
+@LSP_SERVER.feature(NOTEBOOK_DOCUMENT_DID_CHANGE)
+async def did_change_notebook(params: DidChangeNotebookDocumentParams) -> None:
+    """LSP handler for notebookDocument/didChange request."""
+    await _did_change_or_save_notebook(
+        params.notebook_document.uri, run_types=[Run.OnType]
+    )
+
+
+def _group_diagnostics_by_cell(
+    diagnostics: Iterable[Diagnostic],
+) -> Mapping[int, list[Diagnostic]]:
+    """Group diagnostics by cell index.
+
+    The function will return a mapping from cell number to a list of diagnostics for
+    that cell. The mapping will be empty if the diagnostic doesn't contain the cell
+    information.
+    """
+    cell_diagnostics: dict[int, list[Diagnostic]] = {}
+    for diagnostic in diagnostics:
+        cell = cast(DiagnosticData, diagnostic.data).get("cell")
+        if cell is not None:
+            cell_diagnostics.setdefault(cell, []).append(diagnostic)
+    return cell_diagnostics
+
+
+async def _did_change_or_save_notebook(
+    notebook_uri: str, *, run_types: Sequence[Run]
+) -> None:
+    """Handle notebookDocument/didChange and notebookDocument/didSave requests."""
+    notebook_document = LSP_SERVER.workspace.get_notebook_document(
+        notebook_uri=notebook_uri
+    )
+    if notebook_document is None:
+        log_warning(f"No notebook document found for {notebook_uri!r}")
+        return
+
+    document = Document.from_notebook_document(notebook_document)
+    if lint_run(_get_settings_by_document(document.path)) not in run_types:
+        return
+
+    cell_diagnostics = _group_diagnostics_by_cell(await _lint_document_impl(document))
+
+    # Publish diagnostics for every code cell, replacing the previous diagnostics.
+    for cell_idx, cell in enumerate(notebook_document.cells):
+        if cell.kind is not NotebookCellKind.Code:
+            continue
```

Comment:
We need to loop over each code cell here. To understand why, take for example a Notebook containing 2 cells:

1. On first save, both cell contains a single violation which got added using `notebookDocument/didOpen`
2. Suppose we fix the violation in the first cell and saved the notebook. This sends the `notebookDocument/didSave` request. Now, when we run Ruff on the updated document, we'll get diagnostics _only_ for the second cell. But, we need to clear diagnostics from the first cell which is why we need to loop over every cell.

Reply:
Might be worth adding this explanation to the inline comment